statsmodels/sandbox/gam_gsoc2015/gam.py

```diff
@@ -0,0 +1,309 @@
+class Penalty(object):
+    """
+    A class for representing a scalar-value penalty.
+    Parameters
+    wts : array-like
+        A vector of weights that determines the weight of the penalty
+        for each parameter.
+    Notes
+    -----
+    The class has a member called `alpha` that scales the weights.
+    """
+
+    def __init__(self, wts):
+        self.wts = wts
+        self.alpha = 1.
+
+    def func(self, params):
+        """
+        A penalty function on a vector of parameters.
+        Parameters
+        ----------
+        params : array-like
+            A vector of parameters.
+        Returns
+        -------
+        A scalar penaty value; greater values imply greater
+        penalization.
+        """
+        raise NotImplementedError
+
+    def grad(self, params):
+        """
+        The gradient of a penalty function.
+        Parameters
+        ----------
+        params : array-like
+            A vector of parameters
+        Returns
+        -------
+        The gradient of the penalty with respect to each element in
+        `params`.
+        """
+        raise NotImplementedError
+
+
+class GamPenalty(Penalty):
+    
+    def __init__(self, wts=1, alpha=1, cov_der2=None, der2=None):
+        
+        self.wts = wts #should we keep wts????
+        self.alpha = alpha
+        self.cov_der2 = cov_der2
+        self.der2 = der2
+        self.n_samples = der2.shape[0]
+    
+    def func(self, params):
+        ''' 
+        1) params are the coefficients in the regression model
+        2) der2  is the second derivative of the splines basis
+        '''
+        
+        # The second derivative of the estimated regression function
+        f = np.dot(self.der2, params)
+
+        return self.alpha * np.sum(f**2) / self.n_samples
+
+    def grad(self, params):
+        ''' 
+        1) params are the coefficients in the regression model
+        2) der2  is the second derivative of the splines basis
+        3) cov_der2 is obtained as np.dot(der2.T, der2)
+        '''
+                
+        return 2 * self.alpha * np.dot(self.cov_der2, params) / self.n_samples
+    
+    def deriv2(self, params):
+
+        return 2 * self.alpha * self.cov_der2 / self.n_samples
+
+
+
+
+class MultivariateGamPenalty(Penalty):
+    
+    def __init__(self, wts=None, alpha=None, cov_der2=None, der2=None):
+        '''
+        GAM penalty for multivariate regression
+        - cov_der2 is a list of squared matrix of shape (size_base, size_base)
+        - der2 is a list of matrix of shape (n_samples, size_base) 
+        - alpha is a list of doubles. Each one representing the penalty
+          for each function
+        - wts is a list of doubles of the same length of alpha
+        '''
+        
+        assert(len(cov_der2) == len(der2))
+        
+        # the total number of columns in der2 i.e. the len of the params vector
+        self.n_columns = np.sum(d2.shape[1] for d2 in der2)
```

Comment:
naming convention: we use `k` to refer to variables or columns and `n` for observations or rows, i.e. `k_columns`